jwt/validate_test.go

```diff
@@ -134,7 +158,7 @@ func TestGHIssue10(t *testing.T) {
 			if !assert.Error(t, err, "token.Validate should fail") {
 				return
 			}
-			if !assert.False(t, errors.Is(err, jwt.ErrTokenNotYetValid()), `error should be not ErrNotYetValid`) {
+			if !assert.ErrorIs(t, err, jwt.ErrInvalidAudience(fmt.Errorf("")), `error should be ErrInvalidAudience`) {
```

Comment:
Untested, but I _think_ you can just compare it against `jwt.ErrInvalidAudience(nil)`

And if that works, I think you can do this:

```go
func ErrInvalidAudience() ValidationError { // does not take argument
  return &errInvalidAudience{} // no error set
}
```

Reply:
So, are you proposing that I always instantiate the struct where needed so that I can specify the message, while people from the outside have to use the function and aren't required to pass the error?

Reply:
Sorry, that was a fat finger. Let me try this again in my next post

Reply:
What I want to do is to NOT expose the `error` parameter to the users: i.e. want `jwt.ErrInvalidAudience()`, not `jwt.ErrInvalidAudience(fmt.Errorf(...))` or `jwt.ErrInvalidAudience(nil)`

To allow that, this test needs to pass (I'm only showing the struct instantiation so you can see the comparison that I want to make):

```go
// definition
type invalidAudienceError struct {
  err error
}

// I think errors.Is returns true even if you use an empty struct
if !assert.ErrorIs(t, err, jwt.invalidAudience{err: nil}) { }
```

Note that I haven't ran tests with it so I'm not sure if the above holds. if the above doesn't work, the rest of my post is useless. But I think it works.
If the above works, then you should be able to change your code to do this:

```go
type invalidAudienceErr struct {
  err error
}
var errInvalidAudience = &invalidAudienceErr{} // empty err

func ErrInvalidAudience() error {
   return errInvalidAudience
}

// in your test
if !assert.ErrorIs(t, err, jwt.ErrInvalidAudience(), ....) { }
```

As a result:
* Users see ErrInvalidAudience with a simpler interface (if you were a user, wouldn't you wonder why you need to pass an extra argument just to compare the returned error?)
* ErrInvalidAudience() will always return the same instance of the error, and it's more efficient.

Reply:
Also, you may not have to provide `Unwrap()` to `invalidAudienceErr`... I think. Again, unconfirmed.

Reply:
Ok, I understood well. I agree on changing this, the obvious downside is that I need to instantiate the `struct` everywhere, therefore bypassing what I would define a "constructor". Anyway, the tradeoff is worth a simpler and clearer API 🎉

Reply:
Not sure about `instantiate struct everywhere`. I'm obviously not the one writing this code, but at first glance I don't think you would be instantiating any _extra_ errors.

Reply:
I meant that where I was using `ErrInvalidAudience(...)` before in the internal code I now have to write `&invalidAudienceError{...}`. Yea, it's just one place so it's fine, but since this is basically bypassing the official constructor of the `struct` it may be risky in the future if things change. But, as I said, I agree that this makes the API a lot more clearer

Reply:
Oh, I see. You were writing `return ErrInvalidAudience(subErr)`, and you are saying now you have to write `return invalidAudienceError{err: subErr}`.

If that really bothers you, I'm fine wit you impementing `newInvalidAudienceError(error)`, but I don't think it pays off in this particular case because there is only 1 location where this is done. 


---

This is an aside, so I'm definitely not asking you to make this change, but since in Go you can't enforce use of a constructor within the same package, you will need to separate the errors into a another package to really make the constructors enforceable, e.g.

```go
// internal/errors/errors.go (or similar)
type invalidAudienceError { ... }

func NewInvalidAudience(err error) {
   return &invalidAudienceError{err: err}
}

// jwt/validate.go
import "github.com/lestrrat-go/jwx/v2/internal/errors"

var errInvalidAudience = errors.NewInvalidAudience(nil)
func ErrInvalidAudience() ValidationError {
  return errInvalidAudience{}
}

func validateAudience() ... {
   if audienceIsInvalid {
     return errors.NewInvalidAudience(subErr)
   }
}
```

If you don't do this, developers are free to just write `&invalidAudienceErr{}` anywhere they please.

Reply:
>I don't think it pays off in this particular case because there is only 1 location where this is done

Yup, agree. Change done, but I just found out that this code will `panic`:

```go
err := jwt.ErrInvalidAudience()
_ = err.Error()
```

I think that implementing `Error()` to return an empty `string` will solve the issue. What do you think?

Reply:
Ah... While I think for efficiency's sake an empty string works, it becomes a problem when users try to stringify the empty error:

```go
fmt.Printf("%s", jwt.ErrInvalidAudience)
```

so... maybe

```go
func (err invalidAudienceErr) Error() string {
  if err.err == nil {
     return "jwt: invalid audience"
  }
  return err.err.Error()
}
```

or something similar to

```go
func (err invalidAudienceErr) Error() string {
  return fmt.Errorf(`jwt: invalid audience: %s`, err.err)
}
```

Reply:
>it becomes a problem when users try to stringify the empty error

Good point! Maybe we could just instantiate the `struct` with a default error, instead of implementing the `Error()` method ourselves?

Reply:
ATM I think I prefer implementing an `Error()`. I know there are pros/cons there, because
1. it just irks me that there is a glaring panic
2. I don't want to go the route of a full-fledged error for something that is only used a couple of times

Reply:
Done. Last doubt: 

> >Also, you may not have to provide `Unwrap()` to `invalidAudienceErr`
>
> It works, I tried that already. The reason I chose to call that method was that I didn't want an error that is a validation error to wrap another validation error

Since I edited the message while you were posting yours, I'm not sure you noticed the change. Are we good to go, or should I change this?

Reply:
Nah, it doesn't hurt. You can leave it :)